File: test_detection_without_freq.js
#!/usr/bin/env node
/**
 * Test language detection for languages without frequency data in Node.js
 */

const { detectLanguages, getAvailableCodes } = require('./index.js');
const fs = require('fs');
const path = require('path');

async function getLanguagesWithoutFreqData() {
    // Get all available language codes from alphabet data
    const allCodes = await getAvailableCodes();
    
    // Get languages with frequency data
    const freqDir = path.join(__dirname, 'data', 'freq', 'top200');
    const freqCodes = new Set();
    
    if (fs.existsSync(freqDir)) {
        const freqFiles = fs.readdirSync(freqDir).filter(f => f.endsWith('.txt'));
        freqFiles.forEach(file => {
            freqCodes.add(path.basename(file, '.txt'));
        });
    }
    
    // Find languages without frequency data
    const withoutFreq = allCodes.filter(code => !freqCodes.has(code));
    
    console.log(`Total languages with alphabet data: ${allCodes.length}`);
    console.log(`Languages with frequency data: ${freqCodes.size}`);
    console.log(`Languages WITHOUT frequency data: ${withoutFreq.length}`);
    console.log(`\nFirst 20 languages without frequency data: ${withoutFreq.slice(0, 20)}`);
    
    return withoutFreq;
}

async function testDetectionWithoutFreq(testLang, testText, candidateLangs = null) {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`Testing detection for language: ${testLang}`);
    console.log(`Test text: '${testText}'`);
    console.log(`${'='.repeat(60)}`);
    
    if (candidateLangs === null) {
        // Use a small set of candidates including the test language
        candidateLangs = [testLang, 'en', 'es', 'fr', 'de', 'it'];
    }
    
    console.log(`Candidate languages: ${candidateLangs}`);
    
    try {
        const results = detectLanguages(testText, candidateLangs, {}, candidateLangs.length);
        
        console.log(`\nDetection results:`);
        if (results.length > 0) {
            results.forEach((result, i) => {
                console.log(`  ${i + 1}. ${result[0]}: ${result[1].toFixed(4)}`);
            });
        } else {
            console.log("  No languages detected (all scores below threshold)");
        }
        
        // Check if our target language was detected
        const detectedLangs = results.map(r => r[0]);
        if (detectedLangs.includes(testLang)) {
            const rank = detectedLangs.indexOf(testLang) + 1;
            const score = results[detectedLangs.indexOf(testLang)][1];
            console.log(`\n✓ Target language '${testLang}' detected at rank ${rank} with score ${score.toFixed(4)}`);
        } else {
            console.log(`\n✗ Target language '${testLang}' NOT detected`);
        }
        
    } catch (error) {
        console.log(`Error during detection: ${error.message}`);
        console.error(error);
    }
}

async function main() {
    console.log("WorldAlphabets Language Detection Test (Node.js)");
    console.log("Testing detection for languages without frequency data");
    
    // Find languages without frequency data
    const withoutFreq = await getLanguagesWithoutFreqData();
    
    if (withoutFreq.length === 0) {
        console.log("All languages have frequency data!");
        return;
    }
    
    // Test a few specific languages without frequency data
    const testCases = [
        ["ab", "Аҧсуа бызшәа"],   // Abkhazian - Cyrillic script
        ["cop", "ⲧⲙⲛⲧⲣⲙⲛⲕⲏⲙⲉ"],  // Coptic - distinctive script
        ["gez", "ግዕዝ"],           // Ge'ez - Ethiopic script
        ["vai", "ꕙꔤ"],            // Vai - distinctive script
    ];
    
    for (const [langCode, sampleText] of testCases) {
        if (withoutFreq.includes(langCode)) {
            await testDetectionWithoutFreq(langCode, sampleText);
        } else {
            console.log(`\nSkipping ${langCode} - it has frequency data`);
        }
    }
    
    // Test with a broader candidate set for one case
    if (withoutFreq.includes("ab")) {
        console.log(`\n${'='.repeat(60)}`);
        console.log("Testing Abkhazian with broader candidate set");
        console.log(`${'='.repeat(60)}`);
        
        // Include other Cyrillic languages as candidates
        const cyrillicCandidates = ["ab", "ru", "bg", "mk", "sr", "uk", "be"];
        await testDetectionWithoutFreq("ab", "Аҧсуа бызшәа", cyrillicCandidates);
    }
}

main().catch(console.error);

File: test_detection_without_freq.py
#!/usr/bin/env python3
"""
Test language detection for languages without frequency data.
"""

import sys
from pathlib import Path

# Add the src directory to Python path
sys.path.insert(0, str(Path(__file__).parent / "src"))

from worldalphabets import get_available_codes
from worldalphabets.detect import detect_languages

def get_languages_without_freq_data():
    """Find languages that have alphabet data but no frequency data."""
    # Get all available language codes from alphabet data
    all_codes = get_available_codes()
    
    # Get languages with frequency data
    freq_dir = Path("data/freq/top200")
    freq_codes = set()
    if freq_dir.exists():
        for freq_file in freq_dir.glob("*.txt"):
            freq_codes.add(freq_file.stem)
    
    # Find languages without frequency data
    without_freq = [code for code in all_codes if code not in freq_codes]
    
    print(f"Total languages with alphabet data: {len(all_codes)}")
    print(f"Languages with frequency data: {len(freq_codes)}")
    print(f"Languages WITHOUT frequency data: {len(without_freq)}")
    print(f"\nFirst 20 languages without frequency data: {without_freq[:20]}")
    
    return without_freq

def test_detection_without_freq(test_lang, test_text, candidate_langs=None):
    """Test detection for a language without frequency data."""
    print(f"\n{'='*60}")
    print(f"Testing detection for language: {test_lang}")
    print(f"Test text: '{test_text}'")
    print(f"{'='*60}")
    
    if candidate_langs is None:
        # Use a small set of candidates including the test language
        candidate_langs = [test_lang, 'en', 'es', 'fr', 'de', 'it']
    
    print(f"Candidate languages: {candidate_langs}")
    
    try:
        results = detect_languages(
            test_text,
            candidate_langs=candidate_langs,
            topk=len(candidate_langs)
        )
        
        print("\nDetection results:")
        if results:
            for i, (lang, score) in enumerate(results, 1):
                print(f"  {i}. {lang}: {score:.4f}")
        else:
            print("  No languages detected (all scores below threshold)")
            
        # Check if our target language was detected
        detected_langs = [lang for lang, score in results]
        if test_lang in detected_langs:
            rank = detected_langs.index(test_lang) + 1
            score = results[detected_langs.index(test_lang)][1]
            print(f"\n✓ Target language '{test_lang}' detected at rank {rank} with score {score:.4f}")
        else:
            print(f"\n✗ Target language '{test_lang}' NOT detected")
            
    except Exception as e:
        print(f"Error during detection: {e}")
        import traceback
        traceback.print_exc()

def main():
    print("WorldAlphabets Language Detection Test")
    print("Testing detection for languages without frequency data")
    
    # Find languages without frequency data
    without_freq = get_languages_without_freq_data()
    
    if not without_freq:
        print("All languages have frequency data!")
        return
    
    # Test a few specific languages without frequency data
    test_cases = [
        # Pick some languages that are likely to have distinctive text
        ("aa", "Qafar af keena"),  # Afar - if available
        ("ab", "Аҧсуа бызшәа"),   # Abkhazian - Cyrillic script
        ("cop", "ⲧⲙⲛⲧⲣⲙⲛⲕⲏⲙⲉ"),  # Coptic - distinctive script
        ("gez", "ግዕዝ"),           # Ge'ez - Ethiopic script
        ("vai", "ꕙꔤ"),            # Vai - distinctive script
    ]
    
    for lang_code, sample_text in test_cases:
        if lang_code in without_freq:
            test_detection_without_freq(lang_code, sample_text)
        else:
            print(f"\nSkipping {lang_code} - it has frequency data")
    
    # Test with a broader candidate set for one case
    if "ab" in without_freq:
        print(f"\n{'='*60}")
        print("Testing Abkhazian with broader candidate set")
        print(f"{'='*60}")
        
        # Include other Cyrillic languages as candidates
        cyrillic_candidates = ["ab", "ru", "bg", "mk", "sr", "uk", "be"]
        test_detection_without_freq("ab", "Аҧсуа бызшәа", cyrillic_candidates)

if __name__ == "__main__":
    main()
